backend/app/tasks/evaluate.py

```diff
@@ -0,0 +1,188 @@
+            return
+    messages = [{"role": "user", "content": evaluation_item.query}]
+
+    try:
+        if evaluation_item.response is None or evaluation_item.response == "":
+            response, _ = generate_answer_by_autoflow(messages, evaluation_item.chat_engine)
+            if response is None or response == "":
+                raise Exception("Autoflow response is empty")
+
+            response = response.replace('\n', '\\n').replace('\r', '\\r')
+
+            logger.info(f"Got response from autoflow for evaluation item #{evaluation_item_id}, {response}")
+            evaluation_item.response = response
+            logger.info(f"Successfully get response item #{evaluation_item_id}")
+
+            with Session(engine, expire_on_commit=False) as session:
+                session.merge(evaluation_item)
+                session.commit()
+
+        evaluate_task(evaluation_item)
+
+    except Exception as e:
+        logger.error(f"Failed to evaluate item #{evaluation_item_id}, error: {e}")
+        evaluation_item.error_msg = traceback.format_exc()
+        evaluation_item.status = EvaluationStatus.ERROR
+
+        with Session(engine, expire_on_commit=False) as session:
+            session.merge(evaluation_item)
+            session.commit()
+
+
+def evaluate_task(evaluation_item: EvaluationItem):
+    import os
+    # Workaround: set openai api key in the environment variable
+    #
+    # Ragas only support this way to set openai api key
+    # If we set it the metrics, it will show an error that the key is not set
+    # https://github.com/explodinggradients/ragas/issues/1725
+    os.environ['OPENAI_API_KEY'] = settings.EVALUATION_OPENAI_API_KEY
+
+    logger.info(f"Enter evaluate_task with evaluation item #{evaluation_item.id}")
+    ragas_list = [{
+        "user_input": evaluation_item.query,
+        "reference": evaluation_item.reference,
+        "response": evaluation_item.response,
+    }]
+    logger.debug(f"Response data {evaluation_item.response}")
+
+    ragas_dataset = EvaluationDataset.from_list(ragas_list)
+    logger.debug(f"Dataset {ragas_dataset.to_pandas().head()}")
+
+    evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4o"))
+    evaluator_embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings(model="text-embedding-3-large"))
+
+    metrics = [
+        # LLMContextRecall(llm=evaluator_llm),  # retrieved_contexts required
+        FactualCorrectness(llm=evaluator_llm),
+        # Faithfulness(llm=evaluator_llm),  # retrieved_contexts required
+        SemanticSimilarity(embeddings=evaluator_embeddings)
+    ]
+
+    try:
+        eval_result = evaluate(dataset=ragas_dataset, metrics=metrics, raise_exceptions=True, show_progress=False)
+
+        logger.debug(f"eval_result to_pandas")
+        result_list = eval_result.to_pandas().to_dict(orient='records')
+        logger.debug(f"result list {result_list}")
+        if len(result_list) == 1:
```

Comment:
```suggestion
        if len(result_list) == 0:
           raise Exception(f"Item {evaluation_item.id} cannot get evaluation from ragas")
```

Reply:
Nope, the logic here is not correct. It only added **one** evaluation item for ragas at once. So the expectation of the length of the return is exactly 1.

Reply:
`if len(result_list) != 1:` is ok, the main suggestion is to unify the handling of errors in except statements

Reply:
We want to save the error message. Does this code snippet meet your requirements?
```python
if len(result_list) != 1:
    evaluation_task_item.error_msg = f"Item {evaluation_task_item.id} cannot get evaluation from ragas"
    evaluation_task_item.status = EvaluationStatus.ERROR
else:
    logger.debug(f"result {result_list[0]}")
    evaluation_task_item.factual_correctness = result_list[0][FactualCorrectness.name]
    evaluation_task_item.semantic_similarity = result_list[0][SemanticSimilarity.name]
    evaluation_task_item.status = EvaluationStatus.DONE
```

Reply:
I means:

```python
try:
  if len(result_list) != 1:
    raise Exception(f"Item {evaluation_item.id} cannot get evaluation from ragas")

  logger.debug(f"result {result_list[0]}")
  evaluation_task_item.factual_correctness = result_list[0][FactualCorrectness.name] 
  evaluation_task_item.semantic_similarity = result_list[0][SemanticSimilarity.name]
  evaluation_task_item.status = EvaluationStatus.DONE
  ...
except Exception as e:
  logger.error(f"Failed to evaluate item #{evaluation_item.id}, error: {e}")
  evaluation_item.error_msg = traceback.format_exc()
  evaluation_item.status = EvaluationStatus.ERROR
  ...
```

Reply:
Eh! Sorry for my stupid questions. XD
Updated.